text/0225-configuration.md

```diff
@@ -0,0 +1,411 @@
+
+SDKs validating configuration *MUST* fail when they encounter a configuration with an unsupported version. Generally, this means fail when encountering a major version which is not recognized. An SDK might choose to maintain a library of validators / parsers for each major version, and use the configuration version to select and use the correct instance. Differences in minor versions (except pre-1.0 minor versions) SHOULD be acceptable, with the caveat that allowable property additions and removals MAY result in configuration that is different than excepted.
+
+## Trade-offs and mitigations
+
+### Additional method to configure OpenTelemetry
+
+If the implementation suggested in this OTEP goes ahead, users will be presented with another mechanism for configuring OpenTelemetry. This may cause confusion for users who are new to the project. It may be possible to mitigate the confusion by providing users with best practices and documentation.
+
+### Many ways to configure may result in users not knowing what is configured
+
+As there are multiple mechanisms for configuration, it's possible that the active configuration isn't what was expected. This could happen today, and one way it could be mitigated would be by providing a mechanism to list the active OpenTelemetry configuration.
+
+### Errors or difficulty in configuration files
+
+Configuration files provide an opportunity for misconfiguration. A way to mitigate this would be to provide clear messaging and fail quickly when misconfiguration occurs.
+
+## Prior art and alternatives
+
+The working group looked to the OpenTelemetry Collector and OpenTelemetry Operator for inspiration and guidance.
+
+### Alternative schema languages
+
+In choosing to recommend JSON schema, the working group looked at the following options:
+
+* [Cue](https://cuelang.org/) - A promising simpler language to define a schema, the working group decided against CUE because:
+  * Tooling available for validating CUE files in languages outside of Go were limited.
+  * Familiarity and learning curve would create problems for both users and contributors of OpenTelemetry.
+* [Protobuf](https://developers.google.com/protocol-buffers) - With protobuf already used heavily in OpenTelemetry, the format was worth investigating as an option to define the schema. The working group decided against Protobuf because:
+  * Validation errors are the result of serlization errors which can be difficult to interpret.
+  * Limitations in the schema definition language result in poor ergonomics if type safety is to be retained.
+
+## Open questions
+
+### How to handle environment variable / file config overlap?
+
+How does file configuration interact with environment variable configuration when both are present?
+
+* Solution 1: Ignore environment configuration when file configuration is present. Log a warning to the user indicating that multiple configuration modes were detected, but use the file configuration as the source of truth.
```

Comment:
+1 to the simpler solution here.  IMO we should encourage users to switch to using the config file over env vars over time after it is stable given the benefits outlined above.  Layering of config as described below would make it more difficult to reason about what my program is actually being run with.

Reply:
Agreed. Given the proposed ability to use environment variable substitution in the config files, the need for environment variables should diminish over time.

Reply:
Agree. 
Additionally, perhaps give them a helper config file that uses env var substitution in the right places so that they can migrate easily (and still get a warning until they get rid of env variables and move everything to the config file).

Reply:
I've moved this open question from this section to the "Environment variable substitution" https://github.com/open-telemetry/oteps/pull/225/files#diff-bd4bde8aa8b94a5c3332731193eb54f01b2000c861620e7ec6263d382a4b5979R224 with the following text:

>When both configuration file and environment variables are present, the implementation *MUST* ignore environment variables in preference of the configuration file. The support for environment variable substitution in the configuration file gives users a mechanism for migrating away from environment variables in favour of configuration files.